File: python/pl/grpc/echo/echo_server.py
# Copyright (c) 2026 The Authors. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Authors: liubang (it.liubang@gmail.com)
# Created: 2026/06/21

"""Python gRPC echo server implementing all 5 RPCs."""

import re
import sys
import time
from concurrent import futures

import grpc

from proto.echo.echo_pb2 import (
    ChatMessage,
    EchoRequest,
    EchoResponse,
    EchoSummary,
    HealthRequest,
    HealthResponse,
    ServerStreamRequest,
    StreamItem,
)
from proto.echo.echo_pb2_grpc import (
    EchoServiceServicer,
    StreamServiceServicer,
    add_EchoServiceServicer_to_server,
    add_StreamServiceServicer_to_server,
)


ITEMS = [
    "Alpha", "Bravo", "Charlie", "Delta", "Echo",
    "Foxtrot", "Golf", "Hotel", "India", "Juliet",
]


class EchoService(EchoServiceServicer):
    """Unary RPCs: Echo + HealthCheck."""

    def __init__(self, server_id: str):
        self.server_id = server_id
        self.start_time = time.monotonic()

    def Echo(self, request: EchoRequest, context: grpc.ServicerContext) -> EchoResponse:
        now_us = int(time.time() * 1_000_000)
        return EchoResponse(
            message=request.message,
            original_timestamp=request.timestamp_us,
            server_timestamp=now_us,
            server_id=self.server_id,
        )

    def HealthCheck(
        self, request: HealthRequest, context: grpc.ServicerContext
    ) -> HealthResponse:
        uptime = int(time.monotonic() - self.start_time)
        return HealthResponse(
            status=HealthResponse.SERVING,
            server_id=self.server_id,
            version="1.0.0",
            uptime_seconds=uptime,
        )


class StreamService(StreamServiceServicer):
    """Streaming RPCs: ServerStream + ClientStream + Chat."""

    def __init__(self, server_id: str):
        self.server_id = server_id

    def ServerStream(
        self, request: ServerStreamRequest, context: grpc.ServicerContext
    ):
        pattern = None
        if request.pattern:
            try:
                pattern = re.compile(request.pattern, re.IGNORECASE)
            except re.error:
                context.abort(grpc.StatusCode.INVALID_ARGUMENT, "Invalid regex pattern")

        limit = request.max_responses if request.max_responses > 0 else len(ITEMS)
        count = 0
        for i, content in enumerate(ITEMS):
            if count >= limit:
                break
            if pattern and not pattern.search(content):
                continue
            yield StreamItem(index=i, content=content)
            count += 1

    def ClientStream(
        self, request_iterator, context: grpc.ServicerContext
    ) -> EchoSummary:
        messages = []
        for req in request_iterator:
            messages.append(req.message)
        now_us = int(time.time() * 1_000_000)
        return EchoSummary(
            message_count=len(messages),
            messages=messages,
            server_timestamp=now_us,
            server_id=self.server_id,
        )

    def Chat(self, request_iterator, context: grpc.ServicerContext):
        for msg in request_iterator:
            yield msg


def serve(port: int, server_id: str):
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    add_EchoServiceServicer_to_server(EchoService(server_id), server)
    add_StreamServiceServicer_to_server(StreamService(server_id), server)
    server.add_insecure_port(f"0.0.0.0:{port}")
    server.start()
    print(f"[Python EchoServer] Listening on port {port} (id: {server_id})")
    server.wait_for_termination()


if __name__ == "__main__":
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 50051
    server_id = sys.argv[2] if len(sys.argv) > 2 else "python-server"
    serve(port, server_id)

File: java/pl/grpc/echo/EchoServer.java
// Copyright (c) 2026 The Authors. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Authors: liubang (it.liubang@gmail.com)
// Created: 2026/06/21

package pl.grpc.echo;

import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;
import pl.grpc.proto.*;

import java.io.IOException;
import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

/** Java gRPC echo server using generated stubs. */
public class EchoServer {
    private static final Logger logger = Logger.getLogger(EchoServer.class.getName());
    private static final String VERSION = "1.0.0";
    private static final List<String> ITEMS = List.of(
            "Alpha", "Bravo", "Charlie", "Delta", "Echo",
            "Foxtrot", "Golf", "Hotel", "India", "Juliet");

    static class EchoServiceImpl extends EchoServiceGrpc.EchoServiceImplBase {
        private final String serverId;
        private final Instant startTime;

        EchoServiceImpl(String serverId) {
            this.serverId = serverId;
            this.startTime = Instant.now();
        }

        // NOTE: Java timestamps use System.currentTimeMillis()*1000 which has
        // millisecond (not microsecond) granularity. The field type (int64) is
        // compatible but precision is lower than the C++/Go implementations.
        @Override
        public void echo(EchoRequest req, StreamObserver<EchoResponse> observer) {
            long now = System.currentTimeMillis() * 1000;
            observer.onNext(EchoResponse.newBuilder().setMessage(req.getMessage())
                    .setOriginalTimestamp(req.getTimestampUs()).setServerTimestamp(now)
                    .setServerId(serverId).build());
            observer.onCompleted();
        }

        @Override
        public void serverStream(ServerStreamRequest req, StreamObserver<StreamItem> observer) {
            Pattern pattern = null;
            if (!req.getPattern().isEmpty()) {
                try { pattern = Pattern.compile(req.getPattern(), Pattern.CASE_INSENSITIVE); }
                catch (PatternSyntaxException e) {
                    observer.onError(Status.INVALID_ARGUMENT
                            .withDescription("Invalid regex").asRuntimeException());
                    return;
                }
            }
            int limit = req.getMaxResponses() > 0 ? req.getMaxResponses() : ITEMS.size();
            for (int i = 0, count = 0; i < ITEMS.size() && count < limit; i++) {
                if (pattern != null && !pattern.matcher(ITEMS.get(i)).find()) continue;
                observer.onNext(StreamItem.newBuilder().setIndex(i).setContent(ITEMS.get(i)).build());
                count++;
            }
            observer.onCompleted();
        }

        @Override
        public StreamObserver<EchoRequest> clientStream(StreamObserver<EchoSummary> observer) {
            return new StreamObserver<>() {
                final EchoSummary.Builder summary = EchoSummary.newBuilder();
                int count;
                @Override public void onNext(EchoRequest req) { summary.addMessages(req.getMessage()); count++; }
                @Override public void onError(Throwable t) {
                    logger.warning("ClientStream error: " + t);
                    observer.onError(t); // propagate to client
                }
                @Override public void onCompleted() {
                    long now = System.currentTimeMillis() * 1000;
                    summary.setMessageCount(count).setServerTimestamp(now).setServerId(serverId);
                    observer.onNext(summary.build());
                    observer.onCompleted();
                }
            };
        }

        @Override
        public StreamObserver<ChatMessage> chat(StreamObserver<ChatMessage> observer) {
            return new StreamObserver<>() {
                @Override public void onNext(ChatMessage msg) { observer.onNext(msg); }
                @Override public void onError(Throwable t) {
                    logger.warning("Chat error: " + t);
                    observer.onError(t); // propagate to client
                }
                @Override public void onCompleted() { observer.onCompleted(); }
            };
        }

        @Override
        public void healthCheck(HealthRequest req, StreamObserver<HealthResponse> observer) {
            long uptime = Duration.between(startTime, Instant.now()).getSeconds();
            observer.onNext(HealthResponse.newBuilder().setStatus(HealthResponse.Status.SERVING)
                    .setServerId(serverId).setVersion(VERSION).setUptimeSeconds(uptime).build());
            observer.onCompleted();
        }
    }

    private final Server server;

    public EchoServer(int port, String serverId) throws IOException {
        server = ServerBuilder.forPort(port)
                .addService(new EchoServiceImpl(serverId)).build().start();
        logger.info("[Java EchoServer] Listening on port " + port + " (id: " + serverId + ")");
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            try { server.shutdown().awaitTermination(5, TimeUnit.SECONDS); }
            catch (InterruptedException e) { Thread.currentThread().interrupt(); }
        }));
    }

    public void blockUntilShutdown() throws InterruptedException { server.awaitTermination(); }

    public static void main(String[] args) throws IOException, InterruptedException {
        int port = args.length >= 1 ? Integer.parseInt(args[0]) : 50051;
        String serverId = args.length >= 2 ? args[1] : "java-server";
        new EchoServer(port, serverId).blockUntilShutdown();
    }
}
